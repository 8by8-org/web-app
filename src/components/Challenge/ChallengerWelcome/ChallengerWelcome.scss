@import "src/styles/Components";
@import "src/styles/Fonts";
@import "src/styles/Variables";
@import "src/styles/Classes";

.challenger-welcome {
  .link {
    text-decoration: underline;
    font-weight: bolder;
    &:hover {
      cursor: pointer;
    }
  }

  .content-1 {
    display: flex;
    justify-content: center;
    align-items: center;
    flex-direction: column;
    .background {
      width: 100%;
      height: 300px;
    }
    .container {
      height: 0;
      width: 100%;
      display: flex;
      align-items: center;
      justify-content: center;
      .logo {
        position: relative;
        bottom: 170px;
      }
    }
  }

  .content-2 {
    padding: 30px;
    text-align: center;
    .normal-text {
      margin: 30px 0;
      text-align: left;
    }
    .small-text {
      margin-top: 24px;
      text-align: center;
    }
  }

  .content-3 {
    padding: 30px;
    padding-bottom: 40px;
    padding-top: 0px;
    .normal-heading {
      margin-top: 30px;
    }
    .small-text {
      text-align: center;
      margin-top: 24px;
    }
    .gradient-button {
      margin-top: 30px;
    }
  }
}
