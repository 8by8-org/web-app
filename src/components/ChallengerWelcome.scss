@import "./Styles/Components.scss";

.challenger-welcome {
  background-color: $white;
  color: $black;

  button {
    @include button($black);
    margin: 10px 0;
  }

  .top {
    background-image: url("../assets/images/MaskGroup.png");
    display: flex;
    justify-content: center;
    align-items: center;
    width: 100%;
    height: 300px;

    .logo {
      height: 140px;
    }
  }

  .info {
    padding: 20px 40px;
    display: flex;
    flex-direction: column;
    justify-content: center;

    .title {
      font-size: 30px;
      text-align: center;
      margin: 20px 0;
    }

    .login {
      text-align: center;
      .link {
        color: blue;
        text-decoration: underline;
        &:hover {
          cursor: pointer;
        }
      }
    }
  }
}
