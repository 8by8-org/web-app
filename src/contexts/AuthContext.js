import React, { useContext, useEffect, useState } from "react";
import { getUserDatabaseByUID } from "../functions/UserData";
import { auth } from "../firebase";

export function useAuth() {
  return useContext(AuthContext);
}

export const AuthContext = React.createContext();

export function AuthProvider({ children }) {
  const [currentUser, setCurrentUser] = useState(null);
  //keep track of user data in the context to limit calls to firestore
  const [currentUserData, setCurrentUserData] = useState(null);
  const [authLoading, setAuthLoading] = useState(true);
  //error handling
  const [error, setError] = useState(null);
  /*keep track of voter reg form data here so that the flow can be spread out over several
  discrete pages*/
  const [voterRegistrationData, setVoterRegistrationData] = useState({
    send_confirmation_reminder_emails: false,
    dob: "",
    idNumber: "",
    email: "",
    first_registration: false,
    zip: "",
    citizen: false,
    has_state_license: false,
    eighteenPlus: false,
    name_title: "",
    name_first: "",
    name_middle: "",
    name_last: "",
    street: "",
    streetLine2: "",
    unit: "",
    city: "",
    state: "AL",
    has_mailing_address: false,
    mailing_street: "",
    mailing_streetLine2: "",
    mailing_unit: "",
    mailing_city: "",
    mailing_state: "AL",
    mailing_zip: "",
    phone: "Home",
    change_of_name: false,
    prev_name_title: "",
    prev_name_first: "",
    prev_name_middle: "",
    prev_name_last: "",
    change_of_address: false,
    prev_street: "",
    prev_streetLine2: "",
    prev_unit: "",
    prev_city: "",
    prev_state: "AL",
    prev_zip: "",
    opt_in_email: false,
    opt_in_sms: false,
    opt_in_volunteer: false,
    party: "",
    race: "",
  });

  useEffect(() => {
    const unsubscribe = auth.getAuth().onAuthStateChanged((user) => {
      setError(null);
      setCurrentUser(user);
      //refresh voter registration data
      setVoterRegistrationData({
        dob: "",
        idNumber: "",
        email: user ? user.email : "",
        first_registration: false,
        zip: "",
        citizen: false,
        has_state_license: false,
        eighteenPlus: false,
        title: "",
        name_first: "",
        name_middle: "",
        name_last: "",
        suffix: "",
        street: "",
        streetLine2: "",
        unit: "",
        city: "",
        state: "AL",
        diff_mailing_address: false,
        mail_street: "",
        mail_streetLine2: "",
        mail_unit: "",
        mail_city: "",
        mail_state: "AL",
        mail_zip: "",
        change_of_name: false,
        prev_title: "",
        prev_name_first: "",
        prev_name_middle: "",
        prev_name_last: "",
        prev_suffix: "",
        change_of_address: false,
        prev_street: "",
        prev_streetLine2: "",
        prev_unit: "",
        prev_city: "",
        prev_state: "AL",
        prev_zip: "",
        party: "",
        race: "",
      });
      //if a user has been found, get the current user's information from firestore
      if (user) {
        getUserDatabaseByUID(user.uid)
          .then((userData) => {
            setCurrentUserData(userData);
            setAuthLoading(false);
          })
          .catch((e) => {
            //set an error state to the error message, and set authLoading to false
            setError(e.message);
            setAuthLoading(false);
          });
      } else {
        //if no user was found, set currentUserData to null and authLoading to false
        setCurrentUserData(null);
        setAuthLoading(false);
      }
    });
    return unsubscribe;
  }, []);

  const value = {
    currentUser,
    currentUserData,
    setCurrentUserData,
    voterRegistrationData,
    setVoterRegistrationData,
  };
  return (
    <AuthContext.Provider value={value}>
      {!authLoading && children}
    </AuthContext.Provider>
  );
}
