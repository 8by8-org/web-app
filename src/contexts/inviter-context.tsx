import { Inviter } from "@/models/Inviter";
import { createContext, PropsWithChildren, useState } from "react";

export type InviterContextType = {
  inviterInfo:InviterInfo;
}

type InviterInfo = Inviter | null;

//setting the default context to {activeUser:null} allows for testing unwrapped components
export const InviterContext = createContext<InviterContextType>({inviterInfo: null});

export function InviterContextProvider({children}:PropsWithChildren) {
  const [inviterInfo,setInviterInfo] = useState<InviterInfo>(null);
  return <InviterContext.Provider value={{inviterInfo}}>{children}</InviterContext.Provider>
}

