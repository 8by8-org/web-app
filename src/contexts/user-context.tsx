import {createContext, PropsWithChildren, useState } from "react";
import { User } from "@/models/User";
import { UserType } from "@/models/UserType";

export type UserContextType = {
  activeUser:ActiveUser;
}

type ActiveUser = User | null;

//setting the default context to {activeUser:null} allows for testing unwrapped components
export const UserContext = createContext<UserContextType>({activeUser: null});

export function UserContextProvider({children}:PropsWithChildren) {
  const [activeUser,setActiveUser] = useState<ActiveUser>(null);
  return <UserContext.Provider value={{activeUser}}>{children}</UserContext.Provider>
}

