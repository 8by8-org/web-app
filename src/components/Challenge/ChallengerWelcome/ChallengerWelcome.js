import React from "react";
import { useHistory } from "react-router-dom";
import { useAuth } from "../../../contexts/AuthContext";
import "./ChallengerWelcome.scss";
import Top from "../../../assets/images/ChallengerWelcome/Top.png";
import Logo from "../../../assets/logos/white-logo.svg";
import StepOne from "../../../assets/images/ChallengerWelcome/StepOne.png";
import StepTwo from "../../../assets/images/ChallengerWelcome/StepTwo.png";
import StepThree from "../../../assets/images/ChallengerWelcome/StepThree.png";
import StepFour from "../../../assets/images/ChallengerWelcome/StepFour.png";

function ChallengerWelcome() {
  const history = useHistory();
  sessionStorage.setItem("UserType", "Challenger");

  const { currentUser } = useAuth();

  return (
    <div className="challenger-welcome">
      <div className="content-1">
        <img className="background" src={Top} alt="background" />
        <div className="container">
          <img className="logo" src={Logo} alt="8by8 Logo" />
        </div>
      </div>

      <div className="content-2">
        <div className="normal-title">Welcome!</div>
        <p className="normal-text">
          Closing the voter registration gap has to be a community effort, so
          we're asking everyone to join us in taking the #8by8Challenge—register
          8 friends to register to vote in 8 days!
        </p>
        <button
          onClick={() => history.push("/signup")}
          className="gradient-button"
        >
          Get Started
        </button>
        {!currentUser && (
          <p className="small-text">
            Already have an account?{" "}
            <span className="link" onClick={() => history.push("/signin")}>
              Sign in
            </span>
          </p>
        )}
        <p className="link link-2" onClick={() => history.push("/why8by8")}>
          See why others are doing it
        </p>
      </div>

      <div className="content-3">
        <div className="small-title">Here's How it Works</div>

        <div className="normal-heading">1. Sign Up</div>
        <p className="normal-text">
          Sign up with your name and email address to get started.
        </p>
        <img src={StepOne} alt="sign up" className="center-img" />

        <div className="normal-heading">2. Invite your friends</div>
        <p className="normal-text">
          Get 8 friends via social media or messaging apps to join your
          challenge.
        </p>
        <img src={StepTwo} alt="invite your friends" className="center-img" />

        <div className="normal-heading">3. Friends take action</div>
        <p className="normal-text">
          Your friends can support your challenge by taking 1 of 3 actions:
          register to vote, set up election reminders, or take the challenge
          themselves. You’ll earn 1 badge per friend who takes action!
        </p>
        <img src={StepThree} alt="friends take action" className="center-img" />

        <div className="normal-heading">
          4. Win the challenge, get a reward!
        </div>
        <p className="normal-text">
          When all 8 of your friends took action in your challenge within 8
          days, and you win! Then select and enjoy a reward from one of our
          amazing partners.
        </p>
        <img
          src={StepFour}
          alt="earn 8 badges in 8 days"
          className="center-img"
        />

        <button
          onClick={() => history.push("/signup")}
          className="gradient-button"
        >
          Get Started
        </button>
        {!currentUser && (
          <p className="small-text">
            Already have an account?{" "}
            <span className="link" onClick={() => history.push("/signin")}>
              Sign in
            </span>
          </p>
        )}
      </div>
    </div>
  );
}

export default ChallengerWelcome;
