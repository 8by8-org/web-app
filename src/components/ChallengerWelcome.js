import React from "react";
import { useHistory } from "react-router-dom";
import "./ChallengerWelcome.scss";
import logo from "../assets/logos/white-logo.svg";

function ChallengerWelcome() {
  const history = useHistory();

  return (
    <div className="challenger-welcome">
      <div className="top">
        <img className="logo" src={logo} alt="logo"></img>
      </div>
      <div className="info">
        <div className="title">WELCOME!</div>
        <div>
          We know that closing the voter registration gap has to be a community
          effort, so we're asking everyone to join us in taking the #8by8
          challenge and registering 8 of their friends, family, or coworkers to
          vote in 8 days.
        </div>
        <button onClick={() => history.push("/signup")}>Sign up</button>
        <div className="login">
          Already have an account?{" "}
          <span className="link" onClick={() => history.push("/login")}>
            Sign in
          </span>
        </div>
        <div className="title">HERE'S HOW IT WORKS</div>
        <ul>
          <ol>1. Sign up</ol>
          <ol>2. Invite your friends</ol>
          <ol>3. Wait for them to take action</ol>
          <ol>4. Complete Challenge</ol>
        </ul>
        <button onClick={() => history.push("/signup")}>Sign up</button>
      </div>
    </div>
  );
}

export default ChallengerWelcome;
