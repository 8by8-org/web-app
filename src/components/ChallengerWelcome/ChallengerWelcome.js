import React from "react";
import { useHistory } from "react-router-dom";
import { useAuth } from "../../contexts/AuthContext";
import "./ChallengerWelcome.scss";
import Top from "./../../assets/images/ChallengerWelcome/Top.png";
import Logo from "./../../assets/logos/white-logo.svg";
import StepOne from "./../../assets/images/ChallengerWelcome/StepOne.png";
import StepTwo from "./../../assets/images/ChallengerWelcome/StepTwo.png";
import StepThree from "./../../assets/images/ChallengerWelcome/StepThree.png";
import StepFour from "./../../assets/images/ChallengerWelcome/StepFour.png";

function ChallengerWelcome() {
  const history = useHistory();
  sessionStorage.setItem("UserType", "Challenger");

  const { currentUser } = useAuth();

  return (
    <div className="challenger-welcome">
      <div className="content-1">
        <img className="background" src={Top} alt="background" />
        <div className="container">
          <img className="logo" src={Logo} alt="8by8 Logo" />
        </div>
      </div>

      <div className="content-2">
        <div className="normal-title">Welcome!</div>
        <p className="normal-text">
          Closing the voter registration gap has to be a community effort, so
          we're asking everyone to join us in taking the #8by8Challenge—register
          8 friends to register to vote in 8 days!
        </p>
        <button
          onClick={() => history.push("/signup")}
          className="gradient-button"
        >
          Get Started
        </button>
        {!currentUser && (
          <p className="small-text">
            Already have an account?{" "}
            <span className="link" onClick={() => history.push("/signin")}>
              Sign in
            </span>
          </p>
        )}
      </div>

      <div className="content-3">
        <div className="small-title">Here's How it Works</div>

        <div className="normal-heading">1. Sign Up</div>
        <p className="normal-text">
          Sign up with your name and email address to get started.
        </p>
        <img src={StepOne} alt="sign up" className="center-img" />

        <div className="normal-heading">2. Invite your friends</div>
        <p className="normal-text">
          Get 8 friends via social media or messaging apps to join your
          challenge.
        </p>
        <img src={StepTwo} alt="invite your friends" className="center-img" />

        <div className="normal-heading">3. Friends take action</div>
        <p className="normal-text">
          Your friends can support your challenge by taking 1 of 3 actions:
          register to vote, set up election reminders, or take the challenge
          themselves. You’ll earn 1 badge per friend who takes action!
        </p>
        <img src={StepThree} alt="friends take action" className="center-img" />

        <div className="normal-heading">4. Earn 8 badges in 8 days</div>
        <p className="normal-text">
          When you get 8 badges in 8 days, you win the challenge! Most
          importantly, you helped the community move closer to greater AAPI
          representation!
        </p>
        <img
          src={StepFour}
          alt="earn 8 badges in 8 days"
          className="center-img"
        />

        <button
          onClick={() => history.push("/signup")}
          className="gradient-button"
        >
          Get Started
        </button>
        {!currentUser && (
          <p className="small-text">
            Already have an account?{" "}
            <span className="link" onClick={() => history.push("/signin")}>
              Sign in
            </span>
          </p>
        )}
      </div>
    </div>
  );
}

export default ChallengerWelcome;
