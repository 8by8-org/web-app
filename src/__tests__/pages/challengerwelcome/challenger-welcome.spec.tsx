import { render, cleanup, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import ChallengerWelcome from "@/pages/ChallengerWelcome/challenger-welcome";
import { useRouter } from "next/router";
import { UserContext, UserContextProvider } from "@/contexts/user-context";

jest.mock("next/router", () => ({
  useRouter: jest.fn(),
}));
describe("ChallengerWelcome", () => {
  afterEach(cleanup);

  it("Matches DOM Snapshot", () => {
    const { container } = render(<ChallengerWelcome />);
    expect(container).toMatchSnapshot();
  });
  it("Goes to the sign up page when the buttons are clicked", () => {
    const push = jest.fn();
    (useRouter as jest.Mock).mockImplementation(() => ({
      push,
    }));
    const { container } = render(<ChallengerWelcome />);
    const buttons = container.querySelectorAll("button");
    if (buttons[0]) {
      fireEvent.click(buttons[0]);
    }
    if (buttons[1]) {
      fireEvent.click(buttons[1]);
    }
    expect(push).toHaveBeenNthCalledWith(1, "/signup");
    expect(push).toHaveBeenNthCalledWith(2, "/signup");
  });
  it("Goes to the sign in page when the spans are clicked", () => {
    const push = jest.fn();
    (useRouter as jest.Mock).mockImplementation(() => ({
      push,
    }));
    const { container } = render(
      <UserContextProvider>
        <ChallengerWelcome />
      </UserContextProvider>
    );

    const spans = container.querySelectorAll("span");
    if (spans[0]) {
      fireEvent.click(spans[0]);
    }
    if (spans[1]) {
      fireEvent.click(spans[1]);
    }
    expect(push).toHaveBeenNthCalledWith(1, "/signin");
    expect(push).toHaveBeenNthCalledWith(2, "/signin");
  });
  it("Goes to why8by8 page when the paragraph link is clicked", () => {
    const push = jest.fn();
    (useRouter as jest.Mock).mockImplementation(() => ({
      push,
    }));
    const { container } = render(<ChallengerWelcome />);
    const pLink = container.querySelector(".link-2");
    if (pLink) {
      fireEvent.click(pLink);
    }
    expect(push).toHaveBeenCalledWith("/why8by8");
  });
});
