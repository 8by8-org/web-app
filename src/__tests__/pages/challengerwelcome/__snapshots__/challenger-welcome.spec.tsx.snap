// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChallengerWelcome Matches DOM Snapshot 1`] = `
<div>
  <div
    class="challenger-welcome"
  >
    <div
      class="content-1"
    >
      <img
        alt="background"
        class="background"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
      <div
        class="container"
      >
        <img
          alt="8by8 Logo"
          class="logo"
          data-nimg="1"
          decoding="async"
          height="40"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
          style="color: transparent;"
          width="40"
        />
      </div>
    </div>
    <div
      class="content-2"
    >
      <div
        class="normal-title"
      >
        Welcome!
      </div>
      <p
        class="normal-text"
      >
        Closing the voter registration gap has to be a community effort, so we're asking everyone to join us in taking the #8by8Challengeâ€”register 8 friends to register to vote in 8 days!
      </p>
      <button
        class="gradient-button"
      >
        Get Started
      </button>
      <p
        class="small-text"
      >
        Already have an account?
         
        <span
          class="link"
        >
          Sign in
        </span>
      </p>
      <p
        class="link link-2"
      >
        See why others are doing it
      </p>
    </div>
    <div
      class="content-3"
    >
      <div
        class="small-title"
      >
        Here's How it Works
      </div>
      <div
        class="normal-heading"
      >
        1. Sign Up
      </div>
      <p
        class="normal-text"
      >
        Sign up with your name and email address to get started.
      </p>
      <img
        alt="sign up"
        class="center-img"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
      <div
        class="normal-heading"
      >
        2. Invite your friends
      </div>
      <p
        class="normal-text"
      >
        Get 8 friends via social media or messaging apps to join your challenge.
      </p>
      <img
        alt="invite your friends"
        class="center-img"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
      <div
        class="normal-heading"
      >
        3. Friends take action
      </div>
      <p
        class="normal-text"
      >
        Your friends can support your challenge by taking 1 of 3 actions: register to vote, set up election reminders, or take the challenge themselves. You'll earn 1 badge per friend who takes action!
      </p>
      <img
        alt="friends take action"
        class="center-img"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
      <div
        class="normal-heading"
      >
        4. Win the challenge, get a reward!
      </div>
      <p
        class="normal-text"
      >
        When you get 8 badges in 8 days, you win the challenge! Most importantly, you helped the community move closer to greater AAPI representation!
      </p>
      <img
        alt="earn 8 badges in 8 days"
        class="center-img"
        data-nimg="1"
        decoding="async"
        height="40"
        loading="lazy"
        src="/_next/image?url=%2Fimg.jpg&w=96&q=75"
        srcset="/_next/image?url=%2Fimg.jpg&w=48&q=75 1x, /_next/image?url=%2Fimg.jpg&w=96&q=75 2x"
        style="color: transparent;"
        width="40"
      />
      <button
        class="gradient-button"
        id="get-started"
      >
        Get Started
      </button>
      <p
        class="small-text"
      >
        Already have an account?
         
        <span
          class="link"
        >
          Sign in
        </span>
      </p>
    </div>
  </div>
</div>
`;
